File: etc/python/experimental/BaseXClient.py
# -----------------------------------------------------------------------------
#
# This Python module provides methods to connect to and communicate with the
# BaseX Server.
#
# The Constructor of the class expects a hostname, port, username and password
# for the connection. The socket connection will then be established via the
# hostname and the port.
#
# For the execution of commands you need to call the execute() method with the
# database command as argument. The method returns a boolean, indicating if
# the command was successful. The result can be requested with the result()
# method, and the info() method returns additional processing information
# or error output.
#
# -----------------------------------------------------------------------------
# (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
# -----------------------------------------------------------------------------

import hashlib, socket, array

class Session():

  # Constructor, creating a new socket connection.
  def __init__(self, host, port, user, pw):
    # allocate buffer for speeding up communication
    self.__buf = array.array('B', '\0' * 0x1000)
    self.init()

    # create server connection
    global s
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, port))

    # receive timestamp
    ts = self.readString()

    # send username and hashed password/timestamp
    m = hashlib.md5()
    m.update(hashlib.md5(pw).hexdigest())
    m.update(ts)
    s.send(user + '\0' + m.hexdigest() + '\0')

    # evaluate success flag
    if s.recv(1) != '\0':
      raise IOError('Access Denied.')

  # Executes the specified command.
  def execute(self, com):
    # send command to server
    s.send(com + '\0')

    # send command to server and receive result
    self.init()
    self.__result = self.readString()
    self.__info = self.readString()
    return self.read() == 0
  
  # Returns a query object.  
  def query(self, q):
  	return Query(self, q)

  # Returns the result.
  def result(self):
    return self.__result

  # Returns processing information.
  def info(self):
    return self.__info

  # Closes the connection.
  def close(self):
    s.send('exit\0')
    s.close()

  # Initializes the byte transfer.
  def init(self):
    self.__bpos = 0
    self.__bsize = 0

  # Receives a string from the socket.
  def readString(self):
    bf = array.array('B')
    while True:
      b = self.read()
      if b:
        bf.append(b)
      else:
        return bf.tostring()

  # Returns a single byte from the socket.
  def read(self):
    # Cache next bytes
    if self.__bpos == self.__bsize:
      self.__bsize = s.recv_into(self.__buf)
      self.__bpos = 0
    b = self.__buf[self.__bpos]
    self.__bpos += 1
    return b
  
  # Executes the iterative mode of a query.
  def executeIter(self, com):
  	s.send('\0' + com + '\0');
  	return self.check()
  
  # Sends the defined sign.	
  def send(self, sign):
  	s.send(sign)
  
  # Checks the next byte for null or 1.	
  def check(self):
   	return s.recv(1) == '\0'
   
  # Returns the received string. 	
  def res(self):
  	self.init()
  	return self.readString()
     
   
class Query():
	# Constructor, creating a new query object.
	def __init__(self, session, q):
		self.__session = session
		self.__query = q
	
	# Runs the query and returns the success flag.
	def run(self):
		return self.__session.executeIter(self.__query)
	
	# Checks for more parts of the result.	
	def more(self):
		self.__session.send('\0')
		if self.__session.check():
			self.__part = self.__session.res()
			return True
		else:
			self.close()
			return False
	
	# Returns the next part of the result.
	def next(self):
		return self.__part		
	
	# Closes the iterative execution.	
	def close(self):
		self.__session.send('\1')
	
	# Returns the error info.	
	def info(self):
		return self.__session.res()

File: etc/php/experimental/BaseXClient.php
<?php
/* ----------------------------------------------------------------------------
 *
 * This PHP module provides methods to connect to and communicate with the
 * BaseX Server.
 *
 * The Constructor of the class expects a hostname, port, username and password
 * for the connection. The socket connection will then be established via the
 * hostname and the port.
 *
 * For the execution of commands you need to call the execute() method with the
 * database command as argument. The method returns a boolean, indicating if
 * the command was successful. The result can be requested with the result()
 * method, and the info() method returns additional processing information
 * or error output.
 *
 * ----------------------------------------------------------------------------
 *
 * Example:
 *
 * include("BaseXClient.php");
 *
 * try {
 *   // create session
 *   $session = new Session("localhost", 1984, "admin", "admin");
 *
 *   // perform command and show result or error output
 *   if($session->execute("xquery 1 to 10")) {
 *     print $session->result();
 *   } else {
 *     print $session->info();
 *   }
 *
 *  // close session
 *  $session->close();
 *
 * } catch (Exception $e) {
 *   // print exception
 *   print $e->getMessage();
 * }
 *
 * ----------------------------------------------------------------------------
 * (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
 * ----------------------------------------------------------------------------
 */
class Session {
  /* Class variables.*/
  var $socket, $result, $info, $buffer, $bpos, $bsize;

  /* Constructor, creating a new socket connection. */
  function __construct($h, $p, $user, $pw) {
    // create server connection
    $this->socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
    if(!socket_connect($this->socket, $h, $p)) {
      throw new Exception("Can't communicate with server.");
    }

    // receive timestamp
    $ts = $this->readString();

    // send username and hashed password/timestamp
    $md5 = hash("md5", hash("md5", $pw).$ts);
    socket_write($this->socket, "$user\0$md5\0");

    // receives success flag
    if(socket_read($this->socket, 1) != "\0") {
      throw new Exception("Access denied.");
    }
  }

  /* Executes the specified command. */
  public function execute($com) {
    // send command to server
    socket_write($this->socket, "$com\0");

    // receive result
    $this->init();
    $this->result = $this->readString();
    $this->info = $this->readString();

    return $this->read() == "\0";
  }

  /* Returns the result. */
  public function result() {
    return $this->result;
  }

  /* Returns processing information. */
  public function info() {
    return $this->info;
  }

  /* Closes the connection. */
  public function close() {
    socket_write($this->socket, "exit\0");
    socket_close($this->socket);
  }

  /* Initializes the byte transfer */
  private function init() {
    $this->bpos = 0;
    $this->bsize = 0;
  }

  /* Receives a string from the socket. */
  private function readString() {
    $com = "";
    while(($d = $this->read()) != "\0") {
      $com .= $d;
    }
    return $com;
  }

  /* Returns a single byte from the socket. */
  private function read() {
    if($this->bpos == $this->bsize) {
      $this->bsize = socket_recv($this->socket, $this->buffer, 4096, 0);
      $this->bpos = 0;
    }
    return $this->buffer[$this->bpos++];
  }
  
  public function query($q) {
  	return new Query($this, $q);
  }
  
  public function executeIter($cmd) {
  	// send command to server
    socket_write($this->socket, "\0$com\0");
    return $this->check();
  }
  
  public function send($sign) {
  	socket_write($this->socket, $sign);
  }
  
  public function check() {
  	return socket_read($this->socket, 1) == "\0";
  }
  
  public function res() {
  	$this->init();
  	return $this->readString();
  }
}

class Query {
	/* Class variables.*/
 	var $session, $query, $part;
 	
	function __construct($s, $q) {
		$this->session = $s;
		$this->query = $q;
	}
	
	public function run() {
		return $session->executerIter($query);
	}
	
	public function more() {
		$session->send("\0");
		if ($session->check()) {
			$this->part = $session->res();
			return True;
		} else {
			return False;
		}
	}
	
	public function next() {
		return $part;
	}
	
	public function abort() {
		$session->send("\1");
	}
	
	public function info() {
		return $session->res();
	}	
}
?>
